corim/unsignedcorim.go

```diff
@@ -4,104 +4,219 @@
 }
 
-type ID struct {
-	val interface{}
+// NewUnsignedCorim instantiates an empty UnsignedCorim
+func NewUnsignedCorim() *UnsignedCorim {
+	return &UnsignedCorim{}
 }
 
-func (o ID) MarshalCBOR() ([]byte, error) {
-	return em.Marshal(o.val)
+// SetID sets the corim-id in the unsigned-corim-map to the supplied value.  The
+// corim-id can be passed as UUID in string or binary form (i.e., byte array),
+// or as a (non-empty) string
+func (o *UnsignedCorim) SetID(v interface{}) *UnsignedCorim {
+	if o != nil {
+		tagID := swid.NewTagID(v)
+		if tagID == nil {
+			return nil
+		}
+		o.ID = *tagID
+	}
+	return o
 }
 
-func (o *ID) UnmarshalCBOR(data []byte) error {
-	var s string
+// GetID retrieves the corim-id from the unsigned-corim-map as a string
+func (o UnsignedCorim) GetID() string {
+	return o.ID.String()
+}
 
-	if dm.Unmarshal(data, &s) == nil {
-		o.val = s
-		return nil
-	}
+// AddComid appends the CBOR encoded (and appropriately tagged) CoMID to the
+// tags array of the unsigned-corim-map
+func (o *UnsignedCorim) AddComid(c comid.Comid) *UnsignedCorim {
+	if o != nil {
+		if c.Valid() != nil {
+			return nil
+		}
+
+		comidCBOR, err := c.ToCBOR()
+		if err != nil {
+			return nil
+		}
 
-	var u comid.UUID
+		// d9 01fa # tag(506)
+		comidTag := []byte{0xd9, 0x01, 0xfa}
 
-	if dm.Unmarshal(data, &u) == nil {
-		o.val = u
-		return nil
-	}
+		taggedComid := append(comidTag, comidCBOR...)
 
-	return fmt.Errorf("unknown corim-id type (CBOR: %x)", data)
+		o.Tags = append(o.Tags, taggedComid)
+	}
+	return o
 }
 
-type Tag []byte
+// AddCoswid appends the CBOR encoded (and appropriately tagged) CoSWID to the
+// tags array of the unsigned-corim-map
+func (o *UnsignedCorim) AddCoswid(c swid.SoftwareIdentity) *UnsignedCorim {
+	if o != nil {
+		// Currently the swid package doesn't offer an interface
+		// for validating the supplied CoSWID, so -- for now --
+		// we take any input for granted and pass it to the encoder.
+		// See also https://github.com/veraison/swid/issues/23.
 
-type TaggedComid comid.Comid
+		coswidCBOR, err := c.ToCBOR()
+		if err != nil {
+			return nil
+		}
 
-type Locator struct {
-	Href       comid.TaggedURI `cbor:"0,keyasint" json:"href"`
-	Thumbprint swid.HashEntry  `cbor:"1,keyasint,omitempty" json:"thumbprint,omitempty"`
-}
+		// d9 01f9 # tag(505)
+		coswidTag := []byte{0xd9, 0x01, 0xf9}
 
-func NewUnsignedCorim() *UnsignedCorim {
-	return &UnsignedCorim{}
+		taggedCoswid := append(coswidTag, coswidCBOR...)
+
+		o.Tags = append(o.Tags, taggedCoswid)
+	}
+	return o
 }
 
-func (o *UnsignedCorim) SetIDString(s string) *UnsignedCorim {
+// AddDependentRim creates a corim-locator-map from the supplied arguments and
+// appends it to the dependent RIMs in the unsigned-corim-map
+func (o *UnsignedCorim) AddDependentRim(href string, thumbprint *swid.HashEntry) *UnsignedCorim {
 	if o != nil {
-		if s == "" {
-			return nil
+		l := Locator{
+			Href:       comid.TaggedURI(href),
```

Comment:
Just wondering reviewing line 101, does the dependent RIM href be translated to tagged URI from a CoMID package ? Does this look syntactically correct? Perhaps the same or similar API should exist in CoRIM package as well ?

Reply:
yes, this might end up in a `corim/common` package